### PR TITLE
[Bug Fix] Make sure static literal value is properly cast to it's desired type

### PR DESCRIPTION
**Pull request by steffenbrem**
Fixes issue #688

### CODE REVIEW

File: lib/Doctrine/ODM/PHPCR/Query/Builder/ConverterPhpcr.php [hunk 1]
```diff
@@ -332,4 +332,35 @@ protected function walkOperandDynamicField(OperandDynamicField $node)
...
+
+        if ($field = $node->getParent()->getChildOfType(AbstractLeafNode::NT_OPERAND_DYNAMIC)) {
+            if ($field instanceof OperandDynamicField) {
+                $meta = $this->aliasMetadata[$field->getAlias()];
+                $type = $meta->getFieldMapping($field->getField())['type'];
+
+                $typeMapping = array(
+                    'string' => 'string',
+                    'long' => 'integer',
+                    'decimal' => 'string',
```
Comment (steffenbrem) [inline lib/Doctrine/ODM/PHPCR/Query/Builder/ConverterPhpcr.php]: Should this be cast to a string? According to the docs, it should be.

**Comment by lsmith77** [comment]
thanks.

what issue are you trying to fix specifically?

more importantly if we are not embedding data properly into the query shouldn't this be done inside the PHPCR layer?

a failing test case would help.

**Comment by steffenbrem** [comment]
Well, one use case is for example that if your query builder criteria comes from the query parameters of the `Request`, it always is a string. Since in other doctrine manager you can just pass in strings when it expects integers, I would expect the same for the `phpcr-odm`. Otherwise you would need a lot more application level logic to handle proper type casting. I think it good to eliminate this limitation at the query builder level.

For your question about implementing this in the PHPCR layer, I don't see how you can do this, since you need access to the document metadata (to fetch the fields type mapping).

File: lib/Doctrine/ODM/PHPCR/Query/Builder/ConverterPhpcr.php (hunk 1 above)
Comment (dbu) [inline lib/Doctrine/ODM/PHPCR/Query/Builder/ConverterPhpcr.php]: decimal is arbitrary large numbers, and the idea in PHPCR is to use the BCMATH extension for this. and that is based on strings.

**Comment by dbu** [comment]
sounds sane to me. i would love to see a functional test that fails without your change and succeeds with it. that way we also prevent regressions.

you can ignore the styleci complaint, it is not related to your PR.

**Comment by steffenbrem** [comment]
@dbu On it, I have written a test that checks that you can pass in strings for `long` fields. What do you mean by a test that fails without my change? I don't see how I can accomplish that.

**Comment by steffenbrem** [comment]
I have added a functional test.

**Comment by dbu** [comment]
thanks for the fix!
"test that fails without your change": if we comment out the method you added, but keep in your test, that test should fail, proving the code you provide actually fixes a real bug.

**Pull request merged**